### PR TITLE
[Data] Deprecate min_rows_per_file with partitioned parquet writes

### LINKED ISSUE #59485: [Data] `write_parquet` doesn't respect `min_rows_per_file` if you specify partition columns
### What happened + What you expected to happen

If you specify partition columns, Ray Data might produce files with substantially fewer rows than the specified `min_rows_per_file`. The reason that this happens is because `min_rows_per_file` determines the minimum number of rows per write task. If there are multiple partitions within a write task, then the number of rows written to each file might be too small, even though the task as a whole wrote more than the minimum number of rows.

I don't know if there's an easy solution to this. Maybe we can emit a warning.

### Versions / Dependencies

bd412da04b9845101b6777db1d6381a6f4a917b6

### Reproduction script

```python
import os
import ray
import numpy as np
import pandas as pd
import pyarrow.fs

storage_path = "./data"
os.makedirs(storage_path, exist_ok=True)

NUM_ROWS = 100_000_000
YEARS = [2020, 2021, 2022, 2023, 2024]

initial_path = f"{storage_path}/initial_dataset"
partitioned_path = f"{storage_path}/partitioned_dataset"

print(f"Generating dataset with {NUM_ROWS:,} rows...")


def add_columns(batch):
    n = len(batch["id"])
    batch["month"] = np.random.randint(1, 13, size=n)
    batch["year"] = np.random.choice(YEARS, size=n)
    return batch


ds = ray.data.range(NUM_ROWS)
ds = ds.map_batches(add_columns, batch_format="pandas")

print(f"Writing initial dataset to {initial_path}...")
ds.write_parquet(initial_path)
print("Initial dataset written successfully.")

print(f"\nReading dataset from {initial_path}...")
ds_read = ray.data.read_parquet(initial_path)

print(f"Writing partitioned dataset to {partitioned_path}...")
print("Using partition_cols=['month', 'year'] and min_rows_per_file=10,000,000")
ds_read.write_parquet(
    partitioned_path, partition_cols=["month", "year"], min_rows_per_file=10_000_000
)
print("Partitioned dataset written successfully.")

print("\nValidating partitioned output...")
fs = pyarrow.fs.LocalFileSystem()
abs_path = os.path.abspath(partitioned_path)
files = fs.get_file_info(pyarrow.fs.FileSelector(abs_path, recursive=True))
parquet_files = [f.path for f in files if f.path.endswith(".parquet")]

print(f"Found {len(parquet_files)} parquet files")
sample_file = parquet_files[0]
print(f"Reading sample file: {sample_file}")
df = pd.read_parquet(sample_file)
print(f"Number of rows in sample file: {len(df):,}")

print("\nDone!")
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by 400Ping**
## Description

`min_rows_per_file` can't guarantee a minimum file size with `partition_cols`: each block can be split across partition directories, and no single write task necessarily sees all rows of a partition. Per the maintainer decision in [#59485](https://github.com/ray-project/ray/issues/59485), deprecate the combination instead of silently ignoring the hint.

This follows up on [#53930](https://github.com/ray-project/ray/pull/53930), where the combination works when each partition fits in one write task. To avoid a hard behavior break, passing both options now emits a `DeprecationWarning` and retains the existing behavior. Users who repartitioned by the partition columns can remove `min_rows_per_file` without changing the output layout.

## What changed?

- Emit a `DeprecationWarning` when `write_parquet` is called with both non-empty `partition_cols` and `min_rows_per_file`.
- Apply the same warning to the deprecated `num_rows_per_file` alias.
- Keep `partition_cols=[]` allowed, since it doesn't enable partitioned writes.
- Update the `write_parquet` docstring and Ray Data docs with the migration guidance.
- Update the tests to verify the warnings, successful writes, the deprecated alias, and the empty-list edge case.

## Related issues

Related to [#59485](https://github.com/ray-project/ray/issues/59485)

### CODE REVIEW

**Review by AarryaSaraf — COMMENTED** [review]
No objection to the decision to ban the combination — that matches the maintainer call in #59485. Both comments below are about the **migration path**, which I think has two concrete gaps. I verified the behavior locally (on `master` for the current semantics, and with this PR's check applied by hand for the post-change semantics).

**Review by AarryaSaraf — COMMENTED** [review]
Three smaller follow-ups on the remaining files, split out so they can be resolved independently.

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -4472,7 +4472,10 @@ def write_parquet(
                 rows to each file. If the number of rows per block is larger than the
                 specified value, Ray Data writes the number of rows per block to each file.
                 The specified value is a hint, not a strict limit. Ray Data
-                might write more or fewer rows to each file.
+                might write more or fewer rows to each file. This option isn't
+                supported when non-empty ``partition_cols`` are specified; use
```
Comment (AarryaSaraf) [inline python/ray/data/dataset.py]: Would be better to be explicit. Maybe something like 'raises ValueError when used with partition_cols'

File: python/ray/data/_internal/datasource/parquet_datasink.py [hunk 2]
```diff
@@ -198,6 +198,15 @@ def __init__(
         self.max_rows_per_file = max_rows_per_file
         self.partition_cols = partition_cols
 
+        if self.partition_cols and self.min_rows_per_file is not None:
```
Comment (AarryaSaraf) [inline python/ray/data/_internal/datasource/parquet_datasink.py:202]: This is a hard behavior break with no warning release in front of it, and Ray Data already has an established pattern for exactly this situation. _validate_rows_per_file_args in python/ray/data/_internal/util.py retires a rows-per-file argument like this:

warnings.warn(
"num_rows_per_file is deprecated and will be removed in a future release. "
"Use min_rows_per_file instead.",
DeprecationWarning,
stacklevel=3,
)

File: python/ray/data/_internal/datasource/parquet_datasink.py [hunk 3]
```diff
@@ -198,6 +198,15 @@ def __init__(
...
+            raise ValueError(
```
Comment (AarryaSaraf) [inline python/ray/data/_internal/datasource/parquet_datasink.py]: Worth being explicit about what this actually breaks, because it isn't only people misusing the argument. min_rows_per_file genuinely worked when the data was repartitioned by the partition columns first, which is what the previous version of saving-data.rst recommended and demonstrated. So for the users who followed the documented guidance this PR removes it from them with no warning release and no statement that their output won't change.

File: python/ray/data/dataset.py (hunk 1 above)
Comment (400Ping) [reply python/ray/data/dataset.py]: Done

File: python/ray/data/_internal/datasource/parquet_datasink.py (hunk 2 above)
Comment (400Ping) [reply python/ray/data/_internal/datasource/parquet_datasink.py:202]: Done

File: python/ray/data/_internal/datasource/parquet_datasink.py (hunk 3 above)
Comment (400Ping) [reply python/ray/data/_internal/datasource/parquet_datasink.py]: Done

**Comment by 400Ping** [comment]
Hey @AarryaSaraf please take a look when you have time. Sorry for the late fix, I was at Beijing giving a talk in Community Over Code Asia.

File: python/ray/data/_internal/datasource/parquet_datasink.py [hunk 4]
```diff
@@ -198,6 +199,15 @@ def __init__(
         self.max_rows_per_file = max_rows_per_file
         self.partition_cols = partition_cols
 
+        if self.partition_cols and self.min_rows_per_file is not None:
+            warnings.warn(
+                "`min_rows_per_file` with non-empty `partition_cols` is deprecated. "
+                "Use `repartition(keys=partition_cols)` and `max_rows_per_file` "
```
Comment (AarryaSaraf) [inline python/ray/data/_internal/datasource/parquet_datasink.py]: This warning is missing 2 things:
1. `repartition(keys=partition_cols)` is not enough. Either num_blocks or target_num_rows_per_block must be specified. 
2. A deprecation must come with a timeline (typically 6 months) and warning. We would like to say it will be deprecated in February 2027.

**Comment by AarryaSaraf** [comment]
We cannot say 'Closes https://github.com/ray-project/ray/issues/59485' just yet. That will happen when this is deprecated. For now, we must give people some time to adapt their code (typically 6 months)

File: python/ray/data/tests/datasource/test_parquet.py [hunk 5]
```diff
@@ -2052,81 +2052,47 @@ def _pivot_data(batch: pd.DataFrame) -> pd.DataFrame:  # noqa: WPS430
...
-    actual_df["partition_col"] = actual_df["partition_col"].astype(str)
-    expected_df["partition_col"] = expected_df["partition_col"].astype(str)
 
-    # Align column order and compare.
-    actual_df = actual_df[expected_df.columns]
-    pd.testing.assert_frame_equal(actual_df, expected_df, check_dtype=False)
+def test_write_empty_partition_cols_with_min_rows_per_file(
+    tmp_path,
+    ray_start_regular_shared,
+):
```
Comment (AarryaSaraf) [inline python/ray/data/tests/datasource/test_parquet.py:2153]: I believe your intention with this test was to catch a false positive error (that an empty partition cols might cause); however, you do not make any necessary assertion for that.

File: python/ray/data/tests/datasource/test_parquet.py [hunk 6]
```diff
@@ -2052,81 +2052,47 @@ def _pivot_data(batch: pd.DataFrame) -> pd.DataFrame:  # noqa: WPS430
...
-@pytest.mark.parametrize("min_rows_per_file", [5, 10])
-def test_write_partition_cols_with_min_rows_per_file(
+def test_write_partition_cols_with_min_rows_per_file_warns(
     tmp_path,
     ray_start_regular_shared,
-    min_rows_per_file,
-    target_max_block_size_infinite_or_default,
 ):
-    """Test write_parquet with both partition_cols and min_rows_per_file."""
-
```
Comment (AarryaSaraf) [inline python/ray/data/tests/datasource/test_parquet.py:2063]: Given we dont want to remove this behaviour for some time, we should not remove these tests. You can add to these or add a separate test to assert that the warning is appropriately given

**Comment by AarryaSaraf** [comment]
[nit] title should be changed from 'disallow'

**Review by AarryaSaraf — APPROVED** [review]

**Comment by 400Ping** [comment]
cc @bveeramani to do final pass

File: doc/source/data/saving-data.rst [hunk 7]
```diff
@@ -159,14 +159,24 @@ number of output files, configure ``min_rows_per_file``.
 Writing into Partitioned Dataset
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (dstrodtman) [inline doc/source/data/saving-data.rst]: **Heading case.** The style guide asks for [sentence case in every heading](https://docs.ray.io/en/master/ray-contribute/writing-style.html#headings), and task headings in the imperative. This one predates your PR, so it's genuinely optional — but it's inside your hunk and nothing in the tree cross-references it (no `:ref:` targets it and `autosectionlabel` isn't enabled in `doc/source/conf.py`), so it's a safe rename. The replacement is the same character count, which keeps the RST underline valid.

```suggestion
Write into a partitioned dataset
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Review by dstrodtman — COMMENTED** [review]
## Docs style review (style and grammar only)

Reviewed by **Douglas Strodtman** (`@dstrodtman`), Anyscale documentation team. I used Claude Code to assist with this pass, and I read and stand behind every comment below.

**Scope:** `doc/source/data/saving-data.rst` only, checked against the [Ray documentation style guide](https://docs.ray.io/en/master/ray-contribute/writing-style.html) with the Google developer style guide as the fallback. I'm not a code owner here, so **anything technical is the maintainers' call** — @AarryaSaraf has already approved and @bveeramani is doing the final pass, and nothing below should hold that up. Every comment is a click-acceptable suggestion; take them, adapt them, or skip them.

I'd be glad to give the docs side a stamp once the technical review settles. Ping me and I'll re-read.

### Context I checked first

I read the existing threads before writing anything. The February 2027 timeline, the `repartition(keys=..., num_blocks=...)` guidance, and the deprecation-instead-of-hard-break approach all come from @AarryaSaraf's review and are settled — I haven't touched any of those claims. **My suggestions restyle the prose around them and deliberately leave the substance exactly as written.**

### Notes

- Vale is scoped to `doc/source/data/` in the pre-commit hook, so this file *is* gated. It's clean today, but `Google.Headings` and `Google.Colons` are disabled in `.vale.ini`, so heading case is a human check — that's comment 1 below.
- Vale also flags passive voice on lines 164, 166, and 168. My rewrite of that paragraph resolves all three.

### One question for the maintainers (not a style point)

The warning tells readers to use `repartition()` "with the partition columns and an explicit `num_blocks`." `Dataset.repartition` also accepts `target_num_rows_per_block`, and @AarryaSaraf's [earlier comment](https://github.com/ray-project/ray/pull/63368#discussion_r3787187493) phrased the requirement as "either `num_blocks` or `target_num_rows_per_block` must be specified."

Is naming only `num_blocks` here intentional — the recommended path for this case — or should the doc mention both? I left the text as-is in my suggestion either way, since it matches the `DeprecationWarning` in `parquet_datasink.py` and that consistency has its own value. Purely a question.

File: doc/source/data/saving-data.rst [hunk 8]
```diff
@@ -159,14 +159,24 @@ number of output files, configure ``min_rows_per_file``.
...
+recommended to repartition the dataset by the partition columns prior to writing into it.
+This lets you control the file sizes and their number. When the dataset is repartitioned
+by the partition columns, every block should contain all of the rows corresponding to a
+particular partition, meaning that the number of files created should be controlled by
+the repartitioning, with optional limits from the write method, such as
+``max_rows_per_file``. Since every block is written out independently, when writing the
+dataset without prior repartitioning you could potentially get an N number of files per
+partition (where N is the number of blocks in your dataset) with very limited ability to
+control the number of files and their sizes, since every block could potentially carry
+the rows corresponding to any partition.
```
Comment (dstrodtman) [inline doc/source/data/saving-data.rst]: **Paragraph rewrite.** A few style-guide items land on the lines you rewrote here:

- `lets` is on the [avoid list](https://docs.ray.io/en/master/ray-contribute/writing-style.html#focus-on-what-the-reader-does) alongside `allows` and `enables` — the original said "allows you to," so this one swapped one flagged verb for another.
- `prior to` → `before` ([prefer plain words](https://docs.ray.io/en/master/ray-contribute/writing-style.html#prefer-plain-words)).
- "it's recommended to" hedges guidance that should be a [direct instruction](https://docs.ray.io/en/master/ray-contribute/writing-style.html#turn-hedging-into-direct-advice).
- Passive voice on "is repartitioned," "be controlled," and "is written" — these are the three findings Vale reports on this file.
- Missing article in "When writing partitioned dataset."
- "an N number of files" → "N files."
- The guide asks you to [avoid parentheticals](https://docs.ray.io/en/master/ray-contribute/writing-style.html#sentence-structure) and to split long sentences rather than joining clauses.

No technical content changes:

```suggestion
When you write a partitioned dataset using Hive-style, folder-based partitioning,
repartition the dataset by the partition columns first. Repartitioning gives you control
over the number of files and their sizes. After you repartition by the partition columns,
every block holds all the rows for a particular partition, so the repartitioning
determines how many files Ray creates, with optional limits from the write method such as
``max_rows_per_file``. Ray writes every block out independently, so if you write the
dataset without repartitioning first, you can get N files per partition, where N is the
number of blocks in your dataset. In that case you have very limited control over the
number of files and their sizes, because every block can carry rows for any partition.
```

**Optional, and outside what I'd push for:** the style guide asks contributors to [soft-wrap prose](https://docs.ray.io/en/master/ray-contribute/writing-style.html#soft-wrap-prose) — one source line per paragraph — so diffs stay clean. That rule sits under "MyST and Markdown specifics" and the rest of this `.rst` file is hard-wrapped, so I kept your wrap width above rather than making this paragraph the odd one out. Entirely your call.

File: doc/source/data/saving-data.rst [hunk 9]
```diff
@@ -159,14 +159,24 @@ number of output files, configure ``min_rows_per_file``.
...
+
+.. warning::
+    Ray Data deprecates using ``min_rows_per_file`` with non-empty ``partition_cols``
+    and plans to stop supporting this combination after February 2027. Use
+    ``repartition()`` with the partition columns and an explicit ``num_blocks``, and use
+    ``max_rows_per_file`` instead. If the dataset is already repartitioned by the
+    partition columns, removing ``min_rows_per_file`` doesn't change the output layout.
```
Comment (dstrodtman) [inline doc/source/data/saving-data.rst]: **Admonition wording.** Two style-guide points, and one readability nit:

- The contractions rule carves out an explicit exception: ["Avoid contractions in a warning or an error message, where the extra weight of the full form helps."](https://docs.ray.io/en/master/ray-contribute/writing-style.html#use-contractions) So `doesn't` → `does not` here specifically.
- "Ray Data deprecates using X ... and plans to stop supporting" reads as intent rather than behavior. "Support ... ends after February 2027" states the same fact without the hedge, and keeps [present tense](https://docs.ray.io/en/master/ray-contribute/writing-style.html#write-in-present-tense).
- Splitting the first sentence in two makes the deprecation and the end date each scannable on their own.

**The February 2027 date, the `num_blocks` guidance, and the "output layout doesn't change" claim are unchanged** — those are @AarryaSaraf's, and I'm not restating them more strongly than the current text does.

```suggestion
.. warning::
    Using ``min_rows_per_file`` with non-empty ``partition_cols`` is deprecated. Support
    for this combination ends after February 2027. Instead, call ``repartition()`` with
    the partition columns and an explicit ``num_blocks``, and use ``max_rows_per_file``.
    If you already repartition the dataset by the partition columns, removing
    ``min_rows_per_file`` does not change the output layout.
```

File: doc/source/data/saving-data.rst [hunk 10]
```diff
@@ -204,13 +214,12 @@ number of files & their sizes (since every block could potentially carry the row
     #    block – this minimises shuffling during the write.
```
Comment (dstrodtman) [inline doc/source/data/saving-data.rst]: **Code comment, two small things.** `minimises` is British spelling; Ray docs follow US spelling through the Google style guide. The `–` is a Unicode en dash, and the guide asks you to [avoid dashes in prose and restructure instead](https://docs.ray.io/en/master/ray-contribute/writing-style.html#sentence-structure).

This line is context rather than something you changed, so skip it freely if you'd rather keep the diff tight.

```suggestion
    #    block. This minimizes shuffling during the write.
```

**Two more in this snippet, both outside the diff hunks, so there's no suggestion to click:**

- Line 199, `# Sample dataset that we'll partition by ``city`` and ``year``.` — the guide [rules out first-person plural](https://docs.ray.io/en/master/ray-contribute/writing-style.html#avoid-first-person-plural) even in code comments, and that apostrophe is a Unicode right single quote. "Sample dataset to partition by ``city`` and ``year``." fixes both.
- Line 216 ends `city=<value>/year=<value>/....` with four dots. Three is the ellipsis.

Both predate this PR. Mentioning them in case you're in the file anyway.

File: python/ray/data/tests/datasource/test_parquet.py [hunk 11]
```diff
@@ -2070,9 +2071,15 @@ def test_write_partition_cols_with_min_rows_per_file(
...
-    )
+    with pytest.warns(
+        DeprecationWarning,
+        match=r"February 2027.*num_blocks",
+    ):
+        ds.write_parquet(
+            tmp_path,
+            partition_cols=["partition_col"],
+            min_rows_per_file=min_rows_per_file,
+        )
```
Comment (bveeramani) [inline python/ray/data/tests/datasource/test_parquet.py]: We already have two focused tests to validate the warning behavior. Do we also need this assertion here? Feel it makes this test harder to read

File: python/ray/data/tests/datasource/test_parquet.py [hunk 12]
```diff
@@ -2129,6 +2136,44 @@ def test_write_partition_cols_with_min_rows_per_file(
...
 
+def test_write_partition_cols_with_num_rows_per_file_warns(
+    tmp_path,
+    ray_start_regular_shared,
+):
+    ds = ray.data.from_items(
+        [{"partition": index % 2, "value": index} for index in range(10)]
+    )
+
+    with pytest.warns(DeprecationWarning) as recorded_warnings:
```
Comment (bveeramani) [inline python/ray/data/tests/datasource/test_parquet.py]: Here and below -- Any reason we can't use a simple match parameter (`pytest.warns(..., match=...)`) rather then the more verbose assertions below?

File: python/ray/data/tests/datasource/test_parquet.py [hunk 13]
```diff
@@ -2129,6 +2136,44 @@ def test_write_partition_cols_with_min_rows_per_file(
...
+        ds.write_parquet(tmp_path, partition_cols=["partition"], num_rows_per_file=5)
+
+    messages = [str(warning.message) for warning in recorded_warnings]
+    assert any("`num_rows_per_file` is deprecated" in message for message in messages)
+    assert any(
+        "will no longer be supported after February 2027" in message
+        for message in messages
+    )
+    assert ray.data.read_parquet(tmp_path).count() == 10
```
Comment (bveeramani) [inline python/ray/data/tests/datasource/test_parquet.py]: Nit: Here and below -- Don't think you need this assertion here. It adds runtime to the test suite, and it's not necessary for validating the warning behavior

**Review by bveeramani — APPROVED** [review]
LGTM with some minor comments on the tests

**Comment by 400Ping** [comment]
Sorry, I accidentally rebased before fetching the latest upstream changes. I’ve corrected the branch history without changing the PR content.

**Comment by 400Ping** [comment]
cc @bveeramani to merge
